stdlib/public/runtime/ImageInspectionCommon.cpp

```diff
@@ -24,68 +24,71 @@
-  // 1. Resolve the address from a known-good address like swift5_protocols when
-  //    the image is first loaded (in this function);
-  // 1. Resolve the address from a known-good address like swift5_protocols when
-  //    the address is first used (and atomically swap the address back so we
-  //    don't incur the cost of lookupSymbol() each time we need it; or
-  // 3. Introduce an ABI-breaking change so that all binaries are rebuilt and
-  //    start supplying a value for this field.
+/// Adjust the \c baseAddress field of a metadata sections structure.
+///
+/// \param sections A pointer to a valid \c swift::MetadataSections structure.
+///
+/// This function should be called at least once before the structure or its
+/// address is passed to code outside this file to ensure that the structure's
+/// \c baseAddress field correctly points to the base address of the image it
+/// is describing.
+static void fixupMetadataSectionBaseAddress(swift::MetadataSections *sections) {
+  bool fixupNeeded = false;
 
-#ifndef NDEBUG
 #if defined(__ELF__)
   // If the base address was set but the image is an ELF image, it is going to
   // be __dso_handle which is not the value we expect (Dl_info::dli_fbase), so
-  // we need to fix it up. Since the base address is currently unused by the
-  // runtime outside tests, we don't normally do this work.
-  if (auto baseAddress = sections->baseAddress) {
-    swift::SymbolInfo symbolInfo;
-    if (lookupSymbol(baseAddress, &symbolInfo) && symbolInfo.baseAddress) {
-      sections->baseAddress = symbolInfo.baseAddress;
-    }
+  // we need to fix it up.
+  fixupNeeded = true;
+#elif !defined(__MACH__)
+  // For non-ELF, non-Apple platforms, if the base address is nullptr, it
+  // implies that this image was built against an older version of the runtime
+  // that did not capture any value for the base address.
+  auto oldBaseAddress = sections->baseAddress.load(std::memory_order_relaxed);
+  if (!oldBaseAddress) {
+    fixupNeeded = true;
   }
-#endif
 #endif
 
-  return sections->baseAddress;
+  if (fixupNeeded) {
+    // We need to fix up the base address. We'll need a known-good address in
+    // the same image: `sections` itself will work nicely.
+    swift::SymbolInfo symbolInfo;
+    if (lookupSymbol(sections, &symbolInfo) && symbolInfo.baseAddress) {
+        sections->baseAddress.store(symbolInfo.baseAddress,
+                                    std::memory_order_relaxed);
+    }
+  }
 }
 }
 
 SWIFT_RUNTIME_EXPORT
 void swift_addNewDSOImage(swift::MetadataSections *sections) {
-#ifndef NDEBUG
-  record(sections);
+#if 0
+  // Ensure the base address of the sections structure is correct.
+  //
+  // Currently disabled because none of the registration functions below
+  // actually do anything with the baseAddress field. Instead, this function is
+  // called by other individual functions, lower in this file, that yield
+  // metadata section pointers.
```

Comment:
Please change the comment here to explicitly state the function name (`fixupMetadataSectionBaseAddress`).  Otherwise it is confusing - this is definitely an entry point invoked from other modules (that is, *this* function - `swift_addNewDSOImage`).

```suggestion
  // actually do anything with the baseAddress field. Instead, `fixupMetadataSectionBaseAddress` is
  // called by other individual functions, lower in this file, that yield
  // metadata section pointers.
```

Reply:
Will change, sure.